Mathlib/Algebra/Module/PID.lean

```diff
@@ -280,3 +280,72 @@ theorem equiv_free_prod_directSum [h' : Module.Finite R N] :
 #align module.equiv_free_prod_direct_sum Module.equiv_free_prod_directSum
 
 end Module
+
+open FiniteDimensional
+
+variable [Module.Finite R M] (N : Submodule R M)
+
+lemma Submodule.finrank_quotient_add_finrank_of_free [Module.Free R M] :
+    finrank R (M ⧸ N) + finrank R N = finrank R M := by
```

Comment:
I think all of these hold for any commutative domain and can be generalized to `Module.rank` without the finiteness assumption, because you can base exchange to FractionRing F: the sequence remains exact because localization is flat, and the ranks remain unchanged: any R-independent family is F-independent in the localized module, and conversely any F-independent family in the localized module can be lifted to the original module by multiplying each element with an element of R, and it's easy to see the lifted family remains linearly independent.

I feel they're also true over noncommutative domains but I'm not sure.

Reply:
I was thinking about this as well. So I didn't clean this up for review. 
Not sure about noncommutative domains as well.

Reply:
I wouldn't get crazy about noncommutative stuff, at least for now.

Do we have the necessary API for tensor products?

Reply:
For injectivity on the left, I think we have to prove it directly for IsLocalizedModule and translate to TensorProduct if necessary.
For right exactness, we already have [rTensor_mkQ](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/RightExactness.html#rTensor_mkQ) and [TensorProduct.map_surjective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/RightExactness.html#TensorProduct.map_surjective) so I think we should translate the result, rather than doing it again for IsLocalizedModule.

Reply:
I don't think this holds for noncommutative domains in general. I think we need something like left-Ore condition.
I wonder if we should add a class being the union of division rings and domains so that we can generalize results following from the rank nullity theorem without duplicating lemmas.

Reply:
Localization won't work for noncommutative domains in general, yes. But do you have a counterexample to `rank R (M ⧸ N) + rank R N = rank R M`? If R^n doesn't have a linearly independent subset of size > n but R^{n+1} have one of size > n+1 for some n > 0, then `rank R R = 1`, `rank R R^n = n` but `rank R R^{n+1} = aleph0`, which yields a counterexample, but this is [impossible for domains R](https://mathoverflow.net/questions/317422/rings-that-fail-to-satisfy-the-strong-rank-condition).

Reply:
Let's start with the commutative case and think about the noncommutative case later.

BTW there are several conflicts.

Reply:
I don't think `Module.rank` behaves well with non-ore domains. In particular `Module.rank R M` is infinite for every nontrivial `M`. Anyways I will be PRing the commutative version first. 
There isn't an organic way to make this into a type class without bumping into universe issues though. Maybe the following is practical?
```
class HasRankNullity {R} [inst : Ring R] where
  cond := IsDomain R ∨ ∃ h : DivisionRing R, h.toRing = inst
```

Reply:
Continuing the discussion in zulip.